File: apps/mobile/src/types/travelLog.ts
/**
 * 기록 당시의 반려동물 정보 스냅샷.
 * 프로필을 수정하거나 지워도 과거 기록의 이름·사진이 따라 바뀌지 않도록,
 * 화면은 살아있는 Pet을 다시 조회하지 않고 이 스냅샷만 보고 그린다.
 * 동일 개체 판단은 언제나 petId로만 한다(이름은 중복될 수 있고 변경될 수도 있다).
 */
export interface TravelLogPetSnapshot {
  petId: string;
  nameSnapshot: string;
  profileImageSnapshot?: string;
}

/** 사용자가 남긴 순간 하나. 원본 사진과 손글씨·장식이 적용된 완성 이미지를 구분한다. */
export interface TravelLog {
  logId: string;
  /** null이면 특정 여행에 연결되지 않은 개별 기록 */
  tripId: string | null;
  /** 날짜 그룹핑 기준. ISO 날짜 문자열 (YYYY-MM-DD) */
  recordedDate: string;
  /** 방문 시각. 있으면 정렬 1순위로 사용한다. ISO datetime */
  visitedAt: string | null;
  /** 정렬 2순위 (visitedAt이 없을 때). ISO datetime */
  createdAt: string;
  placeId: string | null;
  placeName: string;
  /** 사용자가 업로드한 원본. 재생성/편집 용도로만 쓰고 목록·팝업에는 노출하지 않는다. */
  originalImageUrl: string;
  /** 손글씨·장식이 적용된 완성 이미지. 콜라주·목록·팝업·공유·저장 전부 이 필드를 쓴다. */
  generatedImageUrl: string;
  /** 사용자가 선택적으로 남긴 한 줄. AI가 이미지에 그린 손글씨 문구와는 별개 필드. */
  personalMessage: string | null;
  /** 기록 시점에 함께한 반려동물 스냅샷. DB의 travel_log_pets 행에 대응한다. */
  companions: TravelLogPetSnapshot[];
  /** 날짜 그룹의 대표 로그 여부. 지정 UI는 이번 범위 밖, 데이터만 준비. */
  isRepresentative: boolean;
}

export interface Trip {
  tripId: string;
  title: string;
  /** 장소명 검색이 대상으로 삼는 필드 */
  placeName: string;
  /** ISO 날짜 문자열 (YYYY-MM-DD) */
  startDate: string;
  endDate: string;
  /**
   * 여행 자체에 설정된 동행 반려동물 스냅샷. 로그가 하나도 없어도 존재할 수 있는
   * 독립 여행 단위이므로 하위 로그에서 계산하지 않고 별도로 보관한다(DB의 trip_pets).
   * 하위 로그의 companions와 별개 데이터이므로 어느 한쪽만 임의로 고치지 않는다.
   */
  companions: TravelLogPetSnapshot[];
  logCount: number;
  /** 메인 화면 콜라주용 미리보기. 항상 실제 로그(같은 logId)에서 가져온다. */
  previewLogs: TravelLog[];
}

/** 여행에 연결되지 않은 개별 로그를 월 단위로 묶은 그룹 */
export interface UngroupedLogGroup {
  groupId: string;
  year: number;
  /** 1-12 */
  month: number;
  logCount: number;
  previewLogs: TravelLog[];
}

export type TravelLogListItem =
  | { kind: 'trip'; trip: Trip }
  | { kind: 'ungrouped'; group: UngroupedLogGroup };

export interface DateRange {
  /** ISO 날짜 문자열 (YYYY-MM-DD) */
  start: string;
  end: string;
}

export interface TravelLogFilters {
  placeQuery: string;
  dateRange: DateRange | null;
  petIds: string[];
}

File: apps/api/app/schemas/travel_log.py
"""여행기록 API 스키마.

`visitedAt` 은 **시각**이다 — 리뷰의 `visitedAt` 은 날짜라 같은 이름이지만
형식이 다르다(docs/api/travel-logs.md).

`routeId` 는 DB 의 `travel_logs.route_id` 다. 앱 화면 파라미터 이름이 `tripId`
라서 헷갈리기 쉬운데, **API 는 `routeId` 로 통일**한다.
"""

import uuid
from datetime import date, datetime
from typing import Literal

from pydantic import Field

from app.db.models.enums import GenerationStatus, MomentMood, WritingStyle
from app.schemas.base import APISchema


class TravelLogCompanion(APISchema):
    """함께한 반려동물의 스냅샷.

    `petId` 는 `ON DELETE SET NULL` 이라 프로필을 완전히 지우면 None 이 된다.
    그래도 이름·사진은 남아서 화면이 깨지지 않는다.
    """

    pet_id: uuid.UUID | None
    name_snapshot: str
    profile_image_snapshot: str | None


class TravelLogItem(APISchema):
    """기록 한 건.

    목록·상세·그룹의 `previewLogs` 가 **모두 이 스키마 하나**를 쓴다.
    셋을 따로 두면 필드가 하나 늘 때마다 세 곳을 고쳐야 한다.
    """

    id: uuid.UUID
    route_id: uuid.UUID | None
    place_id: uuid.UUID | None
    place_name_snapshot: str
    recorded_date: date
    visited_at: datetime | None
    original_image_url: str
    generated_image_url: str | None
    writing_style: WritingStyle
    mood: MomentMood | None
    generation_status: GenerationStatus
    personal_message: str | None
    is_representative: bool
    companions: list[TravelLogCompanion]
    created_at: datetime


class TravelLogListResponse(APISchema):
    items: list[TravelLogItem]
    total: int
    limit: int
    offset: int


class TravelLogUpdate(APISchema):
    """보낸 필드만 수정한다.

    `originalImageUrl` 과 `writingStyle` 은 **여기 없다.** 다시 만들려면
    재생성(`POST /travel-logs/{logId}/regenerate`)을 쓴다.

    `petIds` 를 보내면 기존 `travel_log_pets` 를 전부 지우고 새로 저장한다.
    이때 이름·사진 스냅샷도 **현재 시점 값**으로 갱신된다.
    """

    personal_message: str | None = None
    recorded_date: date | None = None
    visited_at: datetime | None = None
    place_id: uuid.UUID | None = None
    place_name: str | None = None
    mood: MomentMood | None = None
    is_representative: bool | None = None
    pet_ids: list[uuid.UUID] | None = None


# ---------------------------------------------------------------------------
# GET /travel-logs/groups
# ---------------------------------------------------------------------------
# 앱 목록 화면은 기록을 여행 단위 / 월 단위로 묶어 보여준다. 앱이 매번 직접
# 묶지 않도록 서버가 그룹을 만들어 내려준다.


class TravelLogRouteSummary(APISchema):
    """여행 단위 그룹의 머리말.

    `companions` 는 **여행 자체의 반려동물**이라 기록의 `companions`
    (`travel_log_pets`)와 출처가 다르다. 어느 한쪽만 고치지 않는다.
    """

    id: uuid.UUID
    title: str
    start_date: date | None
    end_date: date | None
    place_name_snapshot: str | None
    companions: list[TravelLogCompanion]
    log_count: int
    preview_logs: list[TravelLogItem]


class TravelLogMonthSummary(APISchema):
    """여행에 속하지 않은 개별 기록을 연·월로 묶은 것."""

    year: int
    month: int
    log_count: int
    preview_logs: list[TravelLogItem]


class TravelLogRouteGroup(APISchema):
    kind: Literal["route"] = "route"
    route: TravelLogRouteSummary


class TravelLogMonthGroup(APISchema):
    kind: Literal["ungrouped"] = "ungrouped"
    group: TravelLogMonthSummary


class TravelLogGroupsResponse(APISchema):
    items: list[TravelLogRouteGroup | TravelLogMonthGroup] = Field(default_factory=list)
    total: int
    limit: int
    offset: int
